openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst

```diff
@@ -0,0 +1,102 @@
+which can then be used and customized by course authors/admins.
+
+
+Decision
+--------
+
+We propose to implement this as a Django database-backed model with
+historical records, `LearningContextDiscussionConfig`.
+
+While the underlying model will use `LearningContextKeys` (so that this
+could eventually be extended to cover (sites/programs/etc.), the first
+implementing data structure (`data.py`) can be described as:
+
+    @attr.s(frozen=True)
+    class CourseDiscussionConfigData:
+        course_key = attr.ib(type=CourseKey)
+        enabled = attr.ib(type=bool)
+        lti_configuration = attr.ib(type=dict)
+        plugin_configuration = attr.ib(type=dict)
+        provider = attr.ib(type=str)
+
+Discussions can be disabled on a course by setting `enabled=False`;
+if no record exists, the plugin is considered to be disabled by default.
+
+`provider` is the (arbitrary, but stable) key/id for the discussion provider.
+
+`plugin_configuration` is a free-form dictionary of configuration values,
+to be passed to plugin's renderer
+
+`lti_configuration` is a collection of LTI launch configuration data
+(URL, key, secret, etc.), but this data structure should be extracted
+into its own model; see notes below (`Dependent Work`).
+
+Changes to this model are tracked using Django Simple History.
+
+Corressponding getters/setters/helpers will be created to provide access.
+
+
+Dependent Work
+--------------
+
+The configuration model could start by storing the LTI launch
+configuration as a simple dictionary on the model/record, but ultimately
+(initially?), this configuration should be extracted into its own
+table/model (strictly for LTI launch config) and then linked to from
+this table via foreign key.
+
+This abstraction opens the possibility of future work to create a
+generalized, LTI-backed plugin framework.
+
+
+Future Work
+-----------
+
+- `LearningContextDiscussionConfig` could be turned into a generic
+  frontend pluggability table.  The table would have a slot type,
+  (discussions, courseware header, etc.), a plugin type (LTI, iframed
+  without LTI, privileged, etc.), and optionally a foreign key to the
+  LTI table.
+  - I suggest we defer this work until we have (at least in mind) our
+    second implementation, before over-abstracting. Any possible data
+    migration should be simple enough, right?
+
+- Discussion apps/plugins/providers, implemented as Python models, will
+  be available via some discovery mechanism (Python entry points?,
+  hard-coded list?)
+  They will be responsible for:
+  - the provider id/key, referenced by `LearningContextDiscussionConfig`
+  - any relevant renderers, accessors, etc.
+
+- No site/organization-based provider restrictions will be implemented, at this time.
+  This could be added later via an allow/deny-list.
+
+
+Discussion
+----------
+
+- Now that we've simplified this from two models to one, can we give
+  this (main) model a new, simpler name, like `DiscussionsConfig`?
```

Comment:
Yes, please

Reply:
If we're making this discussions specific, then `DiscussionsConfig` sounds right.  I think we can figure out what it's called in the general-purpose plugin configuration sense when we get to that...

Reply:
`DiscussionConfig`?  Singular? 🤔

Reply:
(Oooh, I see that this has a bit of a naming conflict with openedx/core/djangoapps/discussions/apps.py above... hm.  That could be confusing.)

Reply:
I went with `DiscussionsConfiguration` as the main model.